mne/preprocessing/maxwell.py

```diff
@@ -90,25 +103,55 @@ def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3,
 
     # Magnetometers (with coil_class == 1.0) must be scaled by 100 to improve
     # numerical stability as they have different scales than gradiometers
-    coil_scale = np.ones(len(picks))
-    coil_scale[np.array([coil['coil_class'] == 1.0 for coil in coils])] = 100.
+    magscale = 100.
+    coil_scale = np.ones((len(picks), 1))
+    coil_scale[np.array([coil['coil_class'] == 1.0
+                         for coil in coils])] = magscale
 
     # Compute multipolar moment bases
     origin = np.array(origin) / 1000.  # Convert scale from mm to m
     S_in, S_out = _sss_basis(origin, coils, int_order, ext_order)
+
+    # Fine calibration processing (point-like magnetometers and calib. coeffs)
+    if fine_cal_fname is not None:
+        mag_inds = pick_types(raw.info, meg='mag')
+        grad_inds = pick_types(raw.info, meg='grad')
+        grad_info = pick_info(raw.info, grad_inds)
+
+        # Compute point-like mags to incorporate gradiometer imbalance
+        grad_imbalances = np.array([ch['calib_coeff'] for ch in cal_chans
+                                    if ch['ch_type'] == 'grad'])
+        S_in_fine, S_out_fine = _sss_basis_point(origin, grad_info, int_order,
+                                                 ext_order, grad_imbalances)
+
+        # Add point like magnetometer data to internal/external bases.
+        S_in[grad_inds] += S_in_fine
+        S_out[grad_inds] += S_out_fine
+
+        # Scale magnetometers by calibration coefficient
+        mag_calib_coeffs = np.array([ch['calib_coeff'] for ch in cal_chans
+                                     if ch['ch_type'] == 'mag'])
+        S_in[mag_inds] /= mag_calib_coeffs
+        S_out[mag_inds] /= mag_calib_coeffs
+
+    # Normalize internal and external spaces
+    for spc in [S_in, S_out]:
+        spc /= np.sqrt(np.sum(spc * spc, axis=0))[np.newaxis, :]
+
+    # Combine internal and external spaces
     S_tot = np.c_[S_in, S_out]
 
     # Pseudo-inverse of total multipolar moment basis set (Part of Eq. 37)
     pS_tot = pinv(S_tot, cond=1e-15)
     # Compute multipolar moments of (magnetometer scaled) data (Eq. 37)
-    mm = np.dot(pS_tot, data * coil_scale[:, np.newaxis])
+    mm = np.dot(pS_tot, data * coil_scale)
     # Reconstruct data from internal space (Eq. 38)
     recon = np.dot(S_in, mm[:S_in.shape[1], :])
 
     # Return reconstructed raw file object
     raw_sss = _update_sss_info(raw.copy(), origin, int_order, ext_order,
                                data.shape[0])
-    raw_sss._data[picks, :] = recon / coil_scale[:, np.newaxis]
+    raw_sss._data[picks, :] = recon / coil_scale
```

Comment:
to save memory:
```
raw_sss._data[picks] = recon
raw_sss._data[picks] /= coil_scale
```
otherwise a temporary array the same size as `recon` gets created.